### DISCUSSION TITLE
Async Tasks stuck in Queued State

### DISCUSSION THREAD

**Discussion by jaymekb-eb** [apache/airflow#37981] [General]
### Initial Indication of Issue
Async tasks (specifically sensors) began showing as `Queued` in the Airflow UI after the sensor was completed.  

In the logs for the task instance we can see the trigger firing which afaik indicates that the triggerer is done processing.
The logs below are the last lines of the task logs for a `DateTimeSensorAsync`
```
[2024-02-20, 04:04:59 UTC] {temporal.py:69} INFO - sleeping 1 second...
[2024-02-20, 04:05:00 UTC] {temporal.py:72} INFO - yielding event with payload DateTime(2024, 2, 20, 4, 5, 0, tzinfo=Timezone('UTC'))
[2024-02-20, 04:05:00 UTC] {triggerer_job_runner.py:616} INFO - Trigger geo_maxwell_ingestion/scheduled__2024-02-20T00:00:00+00:00/wait_5_minutes/-1/1 (ID 145784) fired: TriggerEvent<DateTime(2024, 2, 20, 4, 5, 0, tzinfo=Timezone('UTC'))>
```
Clearing the task, had no impact and we had to manually mark it as a success.

### Versions & Infra Details
Airflow - 2.6.3
Helm Chart - 1.10.0
Executor - CeleryKubernetesExecutor
Metastore - MySql 8.0.35 
Kubernetes - 1.28 (eks.7)

### Suspected Issue
After lots of digging, we've found a correlation between these tasks getting "stuck" and the metric `executor.queued_tasks` growing.  Typically, the `executor.queued_tasks` list grows when tasks start getting scheduled but as they're completed they get removed from the list.
Below is what the metric looks like typically
![image](https://github.com/apache/airflow/assets/108832400/83adf685-0b70-43b3-a86c-c232cd251f26)

When we start seeing async tasks getting stuck in a queued state, we see the metric look like this
![image](https://github.com/apache/airflow/assets/108832400/769b8a34-99e2-40fb-874d-16889e28b9ee)

So far we've seen our tasks getting stuck in Queued state 3 times. Each time, the issue resolves after restarting the scheduler which wipes out this `executor.queued_tasks` since it's in-memory.

We've put an alert in place to identify when the `executor.queued_tasks` size is growing higher than expected, and we've seen that once this starts happening, it's typically a day or two before our tasks start getting stuck.

### Logs
Once our tasks start getting stuck, the logs for those tasks in the scheduler seem to go through a loop of the scheduler trying to send them to the executor but the executor not picking them up due to either them not being in `self.queued_tasks` or if they are there, then the executor is unable to launch and changes the state to Scheduled. (This is my best understanding from looking through the airflow code, but I'm definitely not an expert and I have not figured out how things end up in the executors queued_tasks list).

Below is a log we are seeing for the tasks when they get stuck.
```
"{
    ""log"": ""2024-02-21T09:49:50.066379137Z stdout F [2024-02-21T09:49:50.066+0000] {base_executor.py:150} ERROR - could not queue task TaskInstanceKey(dag_id='marketing_event_product_catalog_export', task_id='marketing_event_product_catalog_export_sensors_group.eb_event_inventory_summary__sensor', run_id='scheduled__2024-02-20T08:00:00+00:00', try_number=1, map_index=-1)""
}"
```

We have also seen tasks getting stuck with logs like the one below.  From what I can tell, this can happen for async tasks, but the issue is that we keep getting this log for hours. When this happens, the task will briefly go back to Scheduled state.  I believe this is also why the update to fail tasks stuck in Queued isn't working for these because the timestamp (`queued_dttm`) is being updated - I verified the values in the DB are more recent than when the task was initially queued.
```
{
    "log": "2024-02-21T09:53:17.894722992Z stdout F [2024-02-21T09:53:17.894+0000] {kubernetes_executor.py:593} INFO - TaskInstance: <TaskInstance: marketing_event_product_catalog_export.marketing_event_product_catalog_export_sensors_group.eb_event_inventory_summary__sensor scheduled__2024-02-20T08:00:00+00:00 [queued]> found in queued state but was not launched, rescheduling"
}
```

I have a lot more logs I can share, but I'm not exactly sure what is helpful.

### Questions/Requests
I've tried digging through the airflow code for our current version (2.6.3) to see if anything sticks out as the issue, but I'm not finding anything.  I've also looked through the latest airflow version and release notes to see if maybe this is an issue that's been fixed, but no luck there either.

1. Can someone explain how items are added to/removed from the `executor.queued_tasks` and `executor.running` lists?
2. Do you know if this is a known issue that would be resolved if we upgrade airflow?
3. Any recommendations on how I can reproduce this issue so I can file a proper Issue?

**Comment by potiuk** [comment]
First thing to try is to upgrade to newer Airflow. The version you use is already pretty old - there are multiple 100s of bugixes implemented since - and it's next to impossible to authoritatively say if your particular issue is solved by any of those - you might want to go throug changelogs and analyse all the changes yourself and see if your problem resembles any of those - but other than that there is not a single person who will tell you "this is fixed by this change" - it's impossible to be sure or even reasonably convinced about that. 

if you upgrade, and see that it is still there, this issue will get more chance that someone will look at. And there is quite a chance that such issue has -in fact -been fixed and upgrading will **just** fix it.

Now, I guess no-one will even have a way to reproduce it without going back with the version and investing a lot of time - and we are all doing any help and diagnosis here when we see and can help - seeing such an issue reported for version released 7 months ago when we had 7 releases after with ~ 400 or so bugfixes implement does not make it likely for somee to want to spend their free time on it.

So my advice - upgrade, and come back if you still see the issue with more circumstances (or - more happy scenario) report here that upgrade fixed the issue.

**Reply by jaymekb-eb** [reply]
Thanks @potiuk I suspected that may be the response :) 

We do have plans to upgrade soon and we will update this ticket once that is complete.  That said, I'm still very interested to understand better how tasks end up queued for the executor. Are there any docs you could point me to or code references (latest version is fine)?

**Reply by potiuk** [reply]
The best doc is code. There is a nice talk few years ago from Ash on how the Airflow 2.0 scheduler works https://youtu.be/DYC4-xElccE - you should watch it.
